app/components/form_components/synonym_autocomplete/script.js

```diff
@@ -0,0 +1,111 @@
+import accessibleAutocomplete from 'accessible-autocomplete'
+import { nodeListForEach } from 'govuk-frontend/govuk/common'
+
+const $allAutocompleteElements = document.querySelectorAll('[data-module="app-synonym-autocomplete"]')
+const defaultValueOption = component => component.getAttribute('data-default-value') || ''
+
+const setupAutoComplete = (component) => {
+  const selectEl = component.querySelector('select')
+  const selectOptions = Array.apply(null, selectEl.options)
+
+  const options = selectOptions.map(option => {
+    const boost = parseInt(option.getAttribute('data-boost'), 10) || 1
+    let synonyms = option.getAttribute('data-synonyms')
+    synonyms = synonyms ? synonyms.split('|') : []
+    return {
+      name: option.label,
+      synonyms: synonyms,
+      boost: boost
+    }
+  })
+
+  // Returns a weighting for an option based on it's 'closeness' to a query.
+  // The higher the weight, the better the match. Returns 0 if no match.
+  const calculateWeight = (option, q) => {
+    const name = option.name.toLowerCase()
+    const synonyms = option.synonyms.map(s => s.toLowerCase())
+    // Basic whitespace trimmed, could improve.
+    const query = q.toLowerCase().trim()
+
+    const indexOfQuery = name.indexOf(query)
+    // Take the 'earliest matching' synonym.
+    const indexesOfQueryInSynonyms = synonyms.map(s => s.indexOf(query))
+    const earliestIndexOfQueryInSynonym = Math.min(...indexesOfQueryInSynonyms.filter(i => i >= 0))
+
+    // Does the query match the option's name in some way?
+    const nameIsExactMatch = name === query
+    const nameStartsWithQuery = indexOfQuery === 0
+    const wordInNameStartsWithQuery = name.split(' ').filter(w => w.toLowerCase().indexOf(query) === 0).length > 0
+    // Does the query match one of the option's synonym in some way?
+    const synonymIsExactMatch = synonyms.some(s => s === query)
+    const synonymStartsWithQuery = earliestIndexOfQueryInSynonym === 0
+
+    if (nameIsExactMatch) {
```

Comment:
I think we can organise these branches (and the ones below) using a hash map, with the keys computed -something like:

```javascript
const weightings = {
  [nameIsExactMatch]: 100,
  [synonymIsExactMatch]: 75,
  [nameStartsWithQuery]: 60,
  [synonymStartsWithQuery]: 50,
  [wordInNameStartsWithQuery]: 25,
  base: 0,
};
```